### PR TITLE
add a test case for matrix sensing SDP

### PR DESCRIPTION
**Pull request by stephentu**
this tests the recovery of a low rank matrix with gaussian measurements using the nuclear norm minimization SDP.

### CODE REVIEW

File: src/mlpack/tests/lrsdp_test.cpp [hunk 1]
```diff
@@ -118,6 +118,63 @@ BOOST_AUTO_TEST_CASE(Johnson844LovaszThetaSDP)
...
+  const size_t p = b.n_elem;
+
+  assert(A.n_rows == p);
+  assert(A.n_cols == m * m);
+
+  float r = 0.5 + sqrt(0.25 + 2 * p);
+  if (ceil(r) > m + n)
+    r = m + n;
+
+  std::srand(0x0FAE8543);
```
Comment (stephentu) [inline src/mlpack/tests/lrsdp_test.cpp]: I've done the same sin here as in #372. Let me update this with a deterministic starting point.

**Comment by rcurtin** [comment]
Looks good to me.  I should buy you lunch for writing more comprehensive tests for the code I originally wrote (and also making it work for something useful!).  Two comments, neither of which need to be entirely addressed before I merge it:

Can you add the description of the test to the pull request as a comment before the test?  To me, at least, it's non-obvious what the test is doing based on the name; I'm not entirely an expert in the matrix recovery area.

The second thought is that on my i7 machine in debug mode, the test takes 90 seconds (so, probably a few minutes on older machines).  This isn't too terribly long (especially since CFTest can take up to an hour (#354)...), but one of the things I notice while watching it is that the LBFGS iterations converge most of the way to the optimum in a handful of iterations, but will spend hundreds of iterations before the convergence criterion is met.  I wonder, if the convergence tolerance was loosened, would LRSDP run much faster and still provide adequate solutions?  I doubt that I'll have the time to dig too deep into that, but if you are thinking that the solver is taking too long, I bet that's a good place to start looking to accelerate it.

**Comment by stephentu** [comment]
To address your second point, I think taking the changes from #370 that I proposed would help to address this issue. Your recent commit 658962feb431677099cc88d895b4e0b5fa2ae404 doesn't entirely address the issue because, while the function values are not changing much the gradient is all over the place and takes a while to settle down.

**Comment by stephentu** [comment]
nevermind about 658962f I didn't really take a close look it has nothing to do with LBFGS.

**Comment by rcurtin** [comment]
Looks good to me, but in testing I discovered something weird (I'd like to see if you can reproduce it with your (presumably) slightly different system configuration).  Using gcc 4.9.2 with Armadillo 4.450.4, I built with cmake -DDEBUG=OFF -DPROFILE=OFF only to discover that when I ran

```
$ bin/mlpack_test -t LRSDPTest/GaussianMatrixSensingSDP
```

I got a segfault.  After digging deeper I discovered that this seems to be a compiler issue related to the use of `auto`.  I removed the use of `auto` and it passed; then I figured out how to upload gists with a command-line tool in order to show you what I did (the authentication mechanism failed entirely, but I'll leave that for another day):

https://gist.github.com/anonymous/7ba752ac4aa5c7d55b16

So, we can avoid `auto` or we can dig deeper into what may be an Armadillo issue or a gcc issue; your preference.  :)

**Comment by stephentu** [comment]
weird! I get this error when compiling with debugging off

```
(gdb) r
Starting program: /home/stephentu/mlpack/build/bin/mlpack_test -t LRSDPTest/GaussianMatrixSensingSDP
warning: Could not load shared library symbols for linux-vdso.so.1.
Do you need "set solib-search-path" or "set sysroot"?
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
Running 1 test case...

error: Mat::init(): out of memory

Catchpoint 1 (exception thrown), 0x00007ffff6991050 in __cxa_throw () from /usr/lib/libstdc++.so.6
(gdb) bt
#0  0x00007ffff6991050 in __cxa_throw () from /usr/lib/libstdc++.so.6
#1  0x00000000004b6926 in void arma::arma_stop_bad_alloc<char [27]>(char const (&) [27]) [clone .constprop.175] ()
#2  0x00000000006bb4f9 in void arma::op_reshape::apply<arma::subview<double> >(arma::Mat<arma::subview<double>::elem_type>&, arma::Op<arma::subview<double>, arma::op_reshape> const&) ()
#3  0x00000000004dd931 in void arma::op_strans::apply_proxy<arma::Op<arma::subview<double>, arma::op_reshape> >(arma::Mat<arma::Op<arma::subview<double>, arma::op_reshape>::elem_type>&, arma::Op<arma::subview<double>, arma::op_reshape> const&) ()
#4  0x00000000006b8e42 in LRSDPTest::GaussianMatrixSensingSDP::test_method() ()
#5  0x00000000006ba71e in LRSDPTest::GaussianMatrixSensingSDP_invoker() ()
#6  0x00000000004f45b7 in boost::unit_test::ut_detail::callback0_impl_t<boost::unit_test::ut_detail::unused, void (*)()>::invoke() ()
#7  0x00007ffff7422231 in ?? () from /usr/lib/libboost_unit_test_framework.so.1.57.0
#8  0x00007ffff740153e in boost::execution_monitor::catch_signals(boost::unit_test::callback0<int> const&) () from /usr/lib/libboost_unit_test_framework.so.1.57.0
#9  0x00007ffff7401d93 in boost::execution_monitor::execute(boost::unit_test::callback0<int> const&) () from /usr/lib/libboost_unit_test_framework.so.1.57.0
#10 0x00007ffff7422345 in boost::unit_test::unit_test_monitor_t::execute_and_translate(boost::unit_test::test_case const&) () from /usr/lib/libboost_unit_test_framework.so.1.57.0
#11 0x00007ffff7408c9f in boost::unit_test::framework_impl::visit(boost::unit_test::test_case const&) () from /usr/lib/libboost_unit_test_framework.so.1.57.0
#12 0x00007ffff743d023 in boost::unit_test::traverse_test_tree(boost::unit_test::test_suite const&, boost::unit_test::test_tree_visitor&) () from /usr/lib/libboost_unit_test_framework.so.1.57.0
#13 0x00007ffff743d023 in boost::unit_test::traverse_test_tree(boost::unit_test::test_suite const&, boost::unit_test::test_tree_visitor&) () from /usr/lib/libboost_unit_test_framework.so.1.57.0
#14 0x00007ffff74041ea in boost::unit_test::framework::run(unsigned long, bool) () from /usr/lib/libboost_unit_test_framework.so.1.57.0
#15 0x00007ffff741ffd7 in boost::unit_test::unit_test_main(bool (*)(), int, char**) () from /usr/lib/libboost_unit_test_framework.so.1.57.0
#16 0x00007ffff6094040 in __libc_start_main () from /usr/lib/libc.so.6
#17 0x00000000004e9833 in _start ()
```

Not really sure why this is the case. Oh well, I pushed a fix.

**Comment by rcurtin** [comment]
Hm, fascinating.  What version of gcc and what version of Armadillo?  I doubt I'll take the time to do some serious digging into this, but I think it may be a reasonable conclusion that (at least for now) `auto` is unsafe with Armadillo when optimized.

**Comment by stephentu** [comment]
```
stephentu@fender build(testsensing)$ g++ --version
g++ (GCC) 4.9.2 20141224 (prerelease)
Copyright (C) 2014 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

stephentu@fender build(testsensing)$ cat /usr/include/armadillo_bits/arma_version.hpp 
// Copyright (C) 2009-2013 NICTA (www.nicta.com.au)
// Copyright (C) 2009-2013 Conrad Sanderson
// 
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.


//! \addtogroup arma_version
//! @{



#define ARMA_VERSION_MAJOR 3
#define ARMA_VERSION_MINOR 820
#define ARMA_VERSION_PATCH 1
#define ARMA_VERSION_NAME  "Mt Cootha"

// snip
```

**Pull request merged**